BeeSwift/AppDelegate.swift

```diff
@@ -142,4 +126,135 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
             completionHandler([.alert, .sound, .badge])
         }
     }
+    
+    // processing user's response to a notification
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        
+        guard response.actionIdentifier == UNNotificationDefaultActionIdentifier else {
+            completionHandler()
+            return
+        }
+        
+        guard let slug = response.notification.request.getSlug() else {
+            completionHandler()
+            return
+        }
+        
+        print("found slug: \(slug)")
+        
+        self.fetchGoalsIfNeededThenViewMatchingGoalWith(slug: slug)
+        
+        completionHandler()
+    }
+    
+    // MARK: opening goals
+    
+    private func fetchGoalsIfNeededThenViewMatchingGoalWith(slug: String) {
+        if CurrentUserManager.sharedManager.goals.isEmpty || CurrentUserManager.sharedManager.goalsFetchedAt.timeIntervalSinceNow < -3600 {
```

Comment:
it makes sense to have these checks elsewhere

in `CurrentUserManager` itself where the fetches take place?


Also, the slug has been provided. One could `fetchUser` checking that the slug appears in the list of goals there.